[USER32_APITEST] Add scrollbar showing/hiding testcase

dusxmt opened the pull request:
## Purpose

Check [CORE-19669](https://jira.reactos.org/browse/CORE-19669) to make sure that CS_HREDRAW and CS_VREDRAW are respected when the client area of a window changes due to the scrollbar disappearing or re-appearing.

JIRA issue: [ROSTESTS-397](https://jira.reactos.org/browse/ROSTESTS-397)

## Proposed changes

Add a new rostests testcase that performs the above-mentioned check.

It works by creating a window, changing its color any time the WM_SIZE handler detects a change in the window's client area size (it switches through the standard 16 VGA colors available in 16-color mode).

The window has a timer which triggers a state transition in a state machine every 500ms, this state machine examines what happens when it makes the scroll bars disappear and re-appear, and it also tries shrinking and growing the window.

The test creates 4 windows, for each possible combination of CS_VREDRAW and CS_HREDRAW.

Doug-Lyons commented:
General suggestion is the review https://reactos.org/wiki/Coding_Style - Indentation and Line Width.
Item No. 1 is as follows:
```
1. Line width must be at most 100 characters.
```

Doug-Lyons reviewed (approved):
See general comment on line width. Otherwise LGTM. Thank you for your contribution.

dusxmt commented:
> General suggestion is the review https://reactos.org/wiki/Coding_Style - Indentation and Line Width. Item No. 1 is as follows:
> 
> ```
> 1. Line width must be at most 100 characters.
> ```

Okay, I'll go through the rules and update the code, I see that there's several of them that I violate.

Also, one thing about writing tests, I might be over-using the ok() macro a bit, using it at any point where something may fail.  Is there perhaps a different mechanism to indicate that a failure has occurred, but not as a test result?

My current rationale is that if a runtime failure occurs, recording it as a test failure will avoid the situation where the program "doesn't even get to the part where it performs the test" and it reports success to the user.  Instead, a failure is reported, and the user is encouraged to figure out why.

It might also be possible to make the test go faster, I might try to speed it up to 250ms per step instead of 500ms, the main thing I'm worried about is making sure that the window is fully updated by the time we get to examining its contents and seeing if it was redrawn, and 500ms per step seemed like a safe choice.

dusxmt commented:
To help alleviate the concern with the test "outrunning" the display system, I changed it to begin the actual test only once the window is visible, once the window receives WM_PAINT.  I've tested it on an extremely slow Windows 11 VM, where it takes literal seconds for the window to spawn, and it seems to work just fine, it didn't detect an error, which it would if Windows couldn't keep up.

HBelusca commented:
>  Is there perhaps a different mechanism to indicate that a failure has occurred, but not as a test result?

If by that, you mean some sort of "trace" mechanism, there is the `trace()` function available, and maybe `warn()` or `err()` as well.

Doug-Lyons reviewed (approved):
LGTM.

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+    switch (FsmState)
+    {
+        case FSM_STATE_START:
+
```
Optional: I would encourage you to strip the empty lines after all those case labels. I will approve also without that optional sugeestion.

JoachimHenze commented:
> (win32ss/user/ntuser/windc.c:750) err: [0003011C] GetDC() without ReleaseDC()!.

Regarding that: I do see only one place where GetDC() is explicitly calles. And the same func does also have a ReleaseDC() which is hit for all return-paths. Possible that one of the called functions has a bug and forgot a ReleaseDC(), but (maybe aside from identifying which one it is, which may make sense even today) I feel like fixing that could be legit to do in a future PR.

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+static int SmallWidth = 0;
+static int SmallHeight = 0;
+
+static void ColorsCleanup()
```
```suggestion
static void ColorsCleanup(void)
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+    return TRUE;
+}
+
+static void RunTestWindow(const char *ClassName, const char *WindowTitle, UINT ClassStyle)
```
```suggestion
static void RunTestWindow(PCSTR ClassName, PCSTR WindowTitle, UINT ClassStyle)
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+    Class.cbClsExtra    = 0;
+    Class.cbWndExtra    = 0;
+    Class.hInstance     = HINST_THISCOMPONENT;
+    Class.hIcon         = LoadIcon(NULL, IDI_APPLICATION);
+    Class.hCursor       = LoadCursor(NULL, IDC_ARROW);
```
```suggestion
    Class.hIcon         = LoadIconW(NULL, IDI_APPLICATION);
    Class.hCursor       = LoadCursorW(NULL, IDC_ARROW);
```

dusxmt replied:
Should I switch everything to unicode?  The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.

dusxmt replied:
Actually, I don't think selecting the API type is correct here, since the type of the IDI_/IDC_ constants will depend on whether UNICODE is defined or not, whereas the versions with an API suffix require the inputs to be of a specific type.

And since the return value is a handle, I think that the natural thing to do is to use these calls without a suffix, to match the type of the IDI_/IDC_ constants.

Though I guess there is potentially value in using the API suffix, to trigger -Werror in case the the state of the UNICODE macro doesn't match the code.

HBelusca replied:
> Should I switch everything to unicode? The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.

I was going to suggest that, actually (we prefer having the explicit unicode stuff)

dusxmt replied:
> > Should I switch everything to unicode? The only piece of text that I have in this testcase is the window caption and class name, and those are currently ANSI strings.
> 
> I was going to suggest that, actually (we prefer having the explicit unicode stuff)

Okay, I've replaced all of the API calls I could with explicit Unicode variants...  With the exception of these two lines, I've kept them on the default, since they need to be the ANSI variant when compiled under user32_apitest.exe, otherwise it would fail on Werror, but if the user were to switch the compilation mode to Unicode, an explicit ANSI variant would cause Werror on the Unicode build.

If you'd like me to use the explicit ANSI versions, please let me know.

dusxmt replied:
Thanks for confirming that I can use the compilation mode versions of those two functions via Mattermost!

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+#define TEST_CLASS_NAME   "ScrollBarRedraw"
+#define TEST_WINDOW_TITLE "ScrollBarRedraw"
+
+EXTERN_C IMAGE_DOS_HEADER __ImageBase;
+#define HINST_THISCOMPONENT ((HINSTANCE)&__ImageBase)
+
```
```suggestion
```
Overkill, just use `GetModuleHandleW(NULL);` (or even just `NULL`, see below)

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+    Class.lpfnWndProc   = WindowProc;
+    Class.cbClsExtra    = 0;
+    Class.cbWndExtra    = 0;
+    Class.hInstance     = HINST_THISCOMPONENT;
```
```suggestion
    Class.hInstance     = GetModuleHandleW(NULL);
```
or even just `NULL` instead of the function call, should still do the correct thing (in this case).

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+static BOOL HaveHRedraw = FALSE;
+static BOOL HaveVRedraw = FALSE;
+
+typedef enum
```
```suggestion
typedef enum _FSM_STATE
```
to avoid nameless enums in the debugger.

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+{
+    UINT Iter;
+
+    if (TEST_COLOR_COUNT != 16)
```
This is already a hardcoded value, so it would probably be better to use an ASSERT here.

dusxmt replied:
100% Agreed, especially after implementing the other recommended changes as well.

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+{
+    UINT Iter;
+
+    for (Iter = 0; Iter < TEST_COLOR_COUNT; Iter++)
```
```suggestion
    for (Iter = 0; Iter < _countof(ColorBrushes); Iter++)
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+    Colors[12] = RGB(0xFF, 0x00, 0x00);  /* Red */
+    Colors[13] = RGB(0xFF, 0x00, 0xFF);  /* Magenta */
+    Colors[14] = RGB(0xFF, 0xFF, 0x00);  /* Yellow */
+    Colors[15] = RGB(0xFF, 0xFF, 0xFF);  /* White */
```
Why not statically initializing them in the `Colors` array definition line 20 ?

dusxmt replied:
You're completely right!

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            }
+            break;
+
+        case WM_SIZE:
+
+            Width = LOWORD(lParam);
+            Height = HIWORD(lParam);
```
```suggestion
        case WM_SIZE:
        {
            int Width = LOWORD(lParam);
            int Height = HIWORD(lParam);
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+                ClientWidth = Width;
+                ClientHeight = Height;
+            }
+            return 0;
```
```suggestion
            return 0;
        }
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+
+            FsmState = FSM_STATE_START;
+            FsmTimer = 0;
+
+            return 0;
```
```suggestion
            return 0;
        }
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+
+    switch (Message)
+    {
+        case WM_CREATE:
+
```
```suggestion
        case WM_CREATE:
        {
            BOOL BoolRC;
            RECT Rect;

```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+
+static LRESULT CALLBACK WindowProc(HWND Window, UINT Message, WPARAM wParam, LPARAM lParam)
+{
+    RECT Rect;
+    BOOL BoolRC;
+    int Width;
+    int Height;
+
```
```suggestion
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            return 0;
+
+        case WM_PAINT:
+
```
```suggestion
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            {
+                FsmTimer = SetTimer(Window, 1, FSM_STEP_PERIOD_MS, NULL);
+                ok(FsmTimer != 0, "Failed to initialize FSM timer, code: %ld\n", GetLastError());
+
```
```suggestion
```

JoachimHenze replied:
doing a call, and then checking its retval is dealing with the same thing. Upvoting Hermes suggestion therefore.

dusxmt replied:
Whoops, you're right, I missed this one, sorry!

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+                return OnResize(Window, Width, Height);
+
+                ClientWidth = Width;
+                ClientHeight = Height;
```
These two lines are currently dead code since they are _**after**_ the `return` statement; please fix it as you were intended the code to be.

dusxmt replied:
Thank you for pointing this out.  I noticed it as well when I was cleaning up the code today, and it helped me uncover quite an interesting bug in my code:
![image](https://github.com/user-attachments/assets/a1f88231-fb9e-4d47-a6cb-6c80da502c65)

It turns out that WM_ERASEBKGND is issued before WM_SIZE.  This means that switching the color in WM_SIZE produces incorrect results, as the previous color was already used.

This means that I think I'll need to switch to using WM_PAINT instead of the implicit WM_ERASEBKGND handler.

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            return 0;
+
+        case WM_TIMER:
+
```
```suggestion
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            break;
+
+        case WM_NCDESTROY:
+
```
```suggestion
```

HBelusca commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,712 @@
+            return 0;
+
+        case WM_DESTROY:
+
```
```suggestion
```

dusxmt commented:
Thank you for your really thorough feedback, @HBelusca!  I've went ahead and implemented all of the proposed changes.  The one place where I've deviated was that I used ANSI routines instead of Unicode, since the only text strings I have that I feed to the win32 APIs here are the window caption and window class name, and keeping them as ANSI strings just makes everything easier.

I've also switched over from using ok() for everything to making use of skip() when there's a runtime error, this is in alignment with how other testcases seem to be doing it.

dusxmt commented:
> LGTM with one small exception. Please review ReactOS debug output showing the following line and other similar lines: (win32ss/user/ntuser/windc.c:750) err: [0003011C] GetDC() without ReleaseDC()!.

I've looked at this, and I suspect that it's caused by the SetWindowPos() calls, those probably invalidate the DC.  I've moved the ReleaseDC() call to before I call SetScrollInfo() or SetWindowPos(), and it seems to have fixed it.

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,720 @@
+
+            HideVertScrollBar(Window);
+            FsmState = FSM_STATE_VSCR_HIDDEN;
+
```
```suggestion
```

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,720 @@
+                           ps.rcPaint.top,
+                           ps.rcPaint.right,
+                           ps.rcPaint.bottom);
+
```
```suggestion
```

dusxmt replied:
Heh, the ReactOS team really seems to not like vertical spaces :)  I think they give the code "breathing room" and make it easier to read, but it's your project, so I'll respect your wishes.

JoachimHenze replied:
Thx. Vertical space can and shall be used to separate 2 things of different concern. Contrary doing a function-call and then checking its retval is dealing with the same thing semantically. So that should be kept close together ideally. A visual separation is a way to transmit information to the reader, but that works only when being applied selectively. If you try to split up everything (e.g. each letter), then you end up splitting up nothing. In general it is a nice trait for c code that can be watched at the same time, aka fitting the screen, is not too limited. A nice trait especially when dealing with allocating and freeing resources in C. It is not that we would dislike vertical space per se. We do just use it wisely to transport meaning.

dusxmt replied:
And that's completely reasonable!  Thanks for the suggestion and for your insights!

dusxmt commented:
> > (win32ss/user/ntuser/windc.c:750) err: [0003011C] GetDC() without ReleaseDC()!.
> 
> Regarding that: I do see only one place where GetDC() is explicitly called. And the same func does also have a ReleaseDC() which is hit for all return-paths. Possible that one of the implicitly called functions has a bug and forgot a ReleaseDC(), but (maybe aside from identifying which one it is, which may make sense even today) I feel like fixing that could be legit to do in a future PR. Can we identify the function call which does the implicit GetDC()? E.g. by setting a breakpoint within GetDC() and check the callstack.

I found out what was causing this, it's the fact that DestroyWindow() immediately stops the window procedure evaluation and begins evaluating the messages for terminating the window.  In this context, it makes perfect sense why we were getting that warning: The code never got to the point where it would call ReleaseDC() in the final iteration.

It's also possible that the DestroyWindow() call eventually returns instead of stopping execution (I haven't played around with it enough to confirm or deny this), and if that was the case, the old code would've caused a double-free, so it's good that we've fixed it.

I should note that @Doug-Lyons noticed this with the DestroyWindow() function as well.

dusxmt commented:
Okay, I've implemented all suggested fixes, as well as adding a new feature: Reporting that the test is incomplete if the window is unexpectedly closed.  I think we're ready for another round of reviews.

Doug-Lyons commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,722 @@
+
+                if (FsmState != FSM_STATE_END)
+                {
+                    skip("Window was closed before test concluded, FsmState: %d, FSM_STATE_END: %d.\n", FsmState, FSM_STATE_END);
```
Please shorten this line. Thanks.

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,722 @@
+
+    if (!RegisterClassW(&Class))
+    {
+        skip("Failed to register window class \"%ls\", code: %ld\n",
```
**Optional**: Personally I find code like that to become shorter and more readable when `\"` is replaced by just `'` because that doesn't need to be escaped. But we are really at nitpicking level already, I don't want to withheld my approval for things like that. Especially not when it is about test-code and not production code. In case you want to improve on that, there are several hits of that same style.
```suggestion
        skip("Failed to register window class '%ls', code: %ld\n",
```

dusxmt replied:
Alrighty, I'll switch it out.  The reason why I tend to go for double quotes is that "they represent strings in C", and the apostrophe is a common diacritics mark in English text, so it might be jarring to have some text with apostrophes enclosed in single-quotes.

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,722 @@
+
+    if (Window == NULL)
+    {
+        skip("Failed to create window of class \"%ls\", code: %ld\n",
```
```suggestion
        skip("Failed to create window of class '%ls', code: %ld\n",
```

JoachimHenze reviewed (approved):
all in all LGTM, approved. Just nitpicks left.

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,722 @@
+                           NULL,
+                           hInst,
+                           NULL);
+
```
```suggestion
```

JoachimHenze commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,722 @@
+
+                if (FsmState != FSM_STATE_END)
+                {
+                    skip("Window was closed before test concluded, FsmState: %d, FSM_STATE_END: %d.\n", FsmState, FSM_STATE_END);
```
```suggestion
                    skip("Window closed before test concluded, FsmState: %d, FSM_STATE_END: %d\n", FsmState, FSM_STATE_END);
```
for example / as a start

dusxmt replied:
Yup yup, I've taken this suggestion.  I've also split up the formatted arguments onto the second line to fit the text into 100 columns.

dusxmt commented:
Alright, I've addressed all the feedback again, implementing your suggestions, let's see if we can get this merged! :)

dusxmt commented:
I've realized that there was a possibility for the window to be closed before it received WM_PAINT (where the actual test to begins), so I've added a check that reports that tests were skipped if such a condition happens, to avoid false positive results if there ends up being a problem that causes the window to close prematurely (perhaps a resource shortage?)

binarymaster commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,727 @@
+/*
+ * PROJECT:     ReactOS api tests
```
```suggestion
 * PROJECT:     ReactOS API tests
```

dusxmt replied:
Updated, thanks!  Do note however that a bunch of other tests have the word "api" lowercase as well, I happened to copy it from one of those.

binarymaster commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,727 @@
+/*
+ * PROJECT:     ReactOS api tests
+ * LICENSE:     ISC (https://spdx.org/licenses/ISC)
```
Any chance to license under GPL/LGPL/MIT/something we already use? I found no cases of this license mentioned in the headers in our source tree.

dusxmt replied:
Sure!  The MIT license is the most similar one to the ISC license that I like to release my code under, so I've switched it over to that.

binarymaster commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,727 @@
+        return 0;
+    }
+
+    if(!FillRgn(hdc, Region, ColorBrushes[CurrentColor]))
```
```suggestion
    if (!FillRgn(hdc, Region, ColorBrushes[CurrentColor]))
```

dusxmt replied:
Thanks!  I missed this one, thanks for going ahead and committing the fix.

oleg-dubinskiy commented on modules/rostests/apitests/user32/ScrollBarRedraw.c:
```diff
@@ -0,0 +1,727 @@
+/*
+ * PROJECT:     ReactOS API tests
+ * LICENSE:     MIT (https://spdx.org/licenses/MIT)
```
Are we sure we want to license this under MIT? Maybe GPL is better? :thinking: Looking to all other files even in this apitest, they are under GPL. Or should it depend purely on the author's desidion? If so, and we can't change it against the author's choice, then OK. But if we can, then GPL seems better for me, since this code is a part of ReactOS and it's not even a common stuff like e. g. declarations in public ddk/ndk/psdk/xdk headers.

learn-more replied:
I always prefer MIT instead of GPL.
GPL is like covid, it is infectious in a bad way.

oleg-dubinskiy replied:
OK, maybe it's an argument in general. But as we know, the code licensed under MIT can be taken, used and even closed by anybody. Isn't that a significant disadvantage? Moreover, this one single part of our apitests can't be useful as-is like that, because all others are licensed under GPL. I suspect it will even not compile, because it depends on other GPL'ed files. So does it make any other sense in this specific case still?

dusxmt replied:
My reasoning is: Am I willing to go to court to enforce the GPL if we were to find someone or some organization using my code in a manner that violates the GPL?  And would I honestly care?  The answer to both of those questions is no.  I follow the copy-free spirit of the BSDs, anyone should be able to use the code for any purpose they desire, as long as they give credit.

And as for compiling the test on its own, outside of ReactOS, it's as simple as replacing the headers with the following:
```
#include <assert.h>
#include <windows.h>
#include "wine/test.h"
```
and using the test in the standard Wine test suite, or you can very easily roll your own test.h to compile the test standalone.

I don't think the test being licensed under a permissive license should cause any issues.  You're free to use it, and linking it with GPL code is okay - the source code is MIT, and the resulting binary is GPL'd.  If, however, you insist on the code of the test being released under the GPL as well, in order for it to be effective, we would have to arrange the transferal of copyright ownership to the ReactOS project, as you cannot expect me to personally go into court to "fight for this testcase".

oleg-dubinskiy replied:
OK, if so, then keep it under MIT. Your arguments are right.

> And as for compiling the test on its own, outside of ReactOS, it's as simple as replacing the headers with the following:

> #include <assert.h>
#include <windows.h>
#include "wine/test.h"

Indeed. Then maybe it makes sense to add these includes into `#ifdnef __REACTOS__` or some another appropriate define, to make it compilable outside of ReactOS, and place ReactOS required defines into `#else` clause then? Or on the contrary: place into `#ifdef __REACTOS__` ROS includes, and then non-ROS ones into `#else` clause? And at the final, add the note about them: which includes are corresponding to compile in ROS environment and which are not. :thinking:

dusxmt replied:
I don't think adding those ifdefs is necessary.  If someone wants to use the testcase outside of ReactOS, they can put in the effort to replace the user32 apitest include (precomp.h) with whatever is necessary for running it outside of this bundle of tests.  If someone from the Wine project wanted to copy it, they'd very quickly know what they need to change in order to get it running after giving the code a quick glance.

oleg-dubinskiy replied:
OK, then it's not necessarily.

dusxmt replied:
Worst case scenario, they can always hit me up with an email :)  The address is in the copyright notice.  But yeah, I'm glad we were able to reach a consensus!

oleg-dubinskiy reviewed (approved):
Looks OK for me too, except one suggestion:

Doug-Lyons commented:
Thank you, @dusxmt for your contribution to ReactOS.

dusxmt commented:
You're welcome!  I'm glad I was able to help!

And thank you everyone for your very helpful feedback!

HBelusca commented:
> I've also switched over from using ok() for everything to making use of skip() when there's a runtime error, this is in alignment with how other testcases seem to be doing it.

Note that these functions allow to count the number of successful or skipped tests.

dusxmt commented:
> > I've also switched over from using ok() for everything to making use of skip() when there's a runtime error, this is in alignment with how other testcases seem to be doing it.
> 
> Note that these functions allow to count the number of successful or skipped tests.

Yup, and I do think it's important to signal to the testing system that "something out of the ordinary has happened", with the tests suddenly being skipped, as it may indicate a different, unrelated bug that affects the test...  or a freak accident (not enough resources)
